library/src/blas1/rocblas_swap.hpp

```diff
@@ -0,0 +1,47 @@
+/* ************************************************************************
+ * Copyright 2016-2019 Advanced Micro Devices, Inc.
+ * ************************************************************************ */
+#pragma once
+#include "handle.h"
+#include "rocblas.h"
+
+template <typename T>
+__forceinline__ __device__ __host__ void rocblas_swap_vals(T* x, T* y)
+{
+    T tmp = *y;
+    *y    = *x;
+    *x    = tmp;
+}
+
+template <typename T>
+__global__ void rocblas_swap_kernel(rocblas_int n, T* x, rocblas_int incx, T* y, rocblas_int incy)
+{
+    ptrdiff_t tid = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    if(tid < n)
+    {
+        rocblas_swap_vals(x + tid * incx, y + tid * incy);
+    }
+}
+
+template <rocblas_int NB, typename T>
+rocblas_status rocblas_swap_template(
+    rocblas_handle handle, rocblas_int n, T* x, rocblas_int incx, T* y, rocblas_int incy)
+{
+    // Quick return if possible.
+    if(n <= 0)
```

Comment:
There is some inconsistency across the board about n < 0 being an invalid size or success. I am guessing this is because of the Fortran implementation. We should discuss this .

Reply:
> There is some inconsistency across the board about n < 0 being an invalid size or success. I am guessing this is because of the Fortran implementation. We should discuss this .

This is consistent -- if `N <= 0`, then swap is a no-op. Fortran code:
```
      IF (N.LE.0) RETURN
```

`N <= 0` means the input vector is empty, so the result depends on the operation:
```
DOT: 0
ASUM: 0
AMAX: 0
SWAP: no-op
COPY: no-op
AXPY: no-op
etc.
```
Matrix dimensions are handled differently from vectors, and cannot have negative values.

BLAS originally consisted only of BLAS1. This way back in the era of Cray vector processors, which could perform nearly optimally on any problem as long as the level-1 vector operations were optimized.

BLAS2 was added later, then BLAS3, when vector processors were declining and superscalar RISC with multi-level caching required special techniques for BLAS2 and BLAS3 performance, rather than simply composing them in terms of BLAS1 vector operations.

BLAS1 was more permissive with sizes and increments, since there were only one-dimensional vectors, so a negative or zero size can simply be considered a "0-length vector". For matrices, it becomes harder to decide what to do when one of the dimensions is negative, so it is flagged as invalid.